source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -456,32 +478,26 @@ void CacheFilter::encodeCachedResponse() {
       CacheResponseCodeDetails::get().ResponseFromCacheFilter);
 
   // If the filter is encoding, 304 response headers and cached headers are merged in encodeHeaders.
-  // If the filter is decoding, we need to serve response headers from cache directly.
+  // If the filter is decoding, we need to serve cached response headers directly.
   if (filter_state_ == FilterState::DecodeServingFromCache) {
     decoder_callbacks_->encodeHeaders(std::move(lookup_result_->headers_), end_stream,
                                       CacheResponseCodeDetails::get().ResponseFromCacheFilter);
   }
 
+  // TODO(yosrym93): Make sure this is the right place to add the callbacks.
+  decoder_callbacks_->addDownstreamWatermarkCallbacks(*this);
```

Comment:
I would recommend initializing this as early as possible (in setEncoderFilterCallbacks)

Reply:
I assume you really meant setEncoderFilterCallbacks rather than a typo here. Since I'm stepping up on the implementation code, I did get confused on why the registration is only provided available via StreamDecoderFilterCallbacks. What if a pure encoder filter (who doesn't have decoder_callbacks) wants to listen to downstream watermark changes? In other words, there's no way you could call: encoder_callbacks_->addDownstreamWatermarkCallbacks(*this).

Reply:
I think it's just an implementation oversight because no one has needed it before.